### PR TITLE
Feature/condo/doma 4845/address service locations boost

### PR DESCRIPTION
**Pull request by AleX83Xpert**
Ability to pass organization tin to improve suggestions

### CODE REVIEW

File: apps/address-service/domains/common/utils/services/suggest/SuggestionKeystoneApp.js [hunk 1]
```diff
@@ -18,37 +18,55 @@ class SuggestionKeystoneApp {
             next()
         }
 
-        app.get('/suggest', setNoCache, async (req, res, next) => {
+        app.all('/suggest', setNoCache, async (req, res, next) => {
+            if (!['GET', 'POST'].includes(req.method)) {
```
Comment (akarjakin) [inline apps/address-service/domains/common/utils/services/suggest/SuggestionKeystoneApp.js:22]: there is no need to create static array `['GET', 'POST']` on each request, better to move it to a constant

File: apps/address-service/domains/common/utils/services/suggest/SuggestionKeystoneApp.js [hunk 2]
```diff
@@ -18,37 +18,55 @@ class SuggestionKeystoneApp {
...
+            }
+
+            const reqQuery = get(req, 'query', {})
+            const reqBody = get(req, 'body', {})
+
+            /**
+             * @param {string} param
+             * @param {*} [def]
+             */
+            const getParam = (param, def) => get(reqBody, param, get(reqQuery, param, def))
```
Comment (akarjakin) [inline apps/address-service/domains/common/utils/services/suggest/SuggestionKeystoneApp.js:33]: def -> defaultValue

File: apps/address-service/domains/common/utils/services/suggest/SuggestionKeystoneApp.js [hunk 3]
```diff
@@ -18,37 +18,55 @@ class SuggestionKeystoneApp {
...
              */
-            const context = String(get(req, ['query', 'context'], ''))
+            const context = String(getParam('context', ''))
 
             /**
              * Number of results to return
              * @type {number|NaN}
              */
-            const count = Number(get(req, ['query', 'count'], undefined))
+            const count = Number(getParam('count'))
```
Comment (akarjakin) [inline apps/address-service/domains/common/utils/services/suggest/SuggestionKeystoneApp.js:64]: Number(undefined) -> NaN

File: apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js [hunk 4]
```diff
@@ -188,18 +186,73 @@ class DadataSuggestionProvider extends AbstractSuggestionProvider {
...
+    }
+
+    /**
+     * @param {string} tin
+     * @returns {Promise<*|null>}
+     */
+    async getOrganization (tin) {
+        const redis = getRedisClient('organization', 'tin')
+        const cached = await redis.get(tin)
+        if (cached) {
```
Comment (akarjakin) [inline apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js:202]: empty line for readability

File: apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js [hunk 5]
```diff
@@ -188,18 +186,73 @@ class DadataSuggestionProvider extends AbstractSuggestionProvider {
...
+        }
+
+        const result = await this.callToDadata(`${this.url}/findById/party`, { query: tin })
+
+        if (result) {
+            const data = get(result, ['suggestions', 0], null)
+            if (data) {
+                await redis.set(tin, JSON.stringify(data), 'EX', ORGANIZATION_TIN_CACHE_TTL)
+            }
+            return data
```
Comment (akarjakin) [inline apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js:213]: empty line for readability

File: apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js [hunk 6]
```diff
@@ -188,18 +186,73 @@ class DadataSuggestionProvider extends AbstractSuggestionProvider {
...
+        const cached = await redis.get(tin)
+        if (cached) {
+            return JSON.parse(cached)
+        }
+
+        const result = await this.callToDadata(`${this.url}/findById/party`, { query: tin })
+
+        if (result) {
+            const data = get(result, ['suggestions', 0], null)
+            if (data) {
```
Comment (akarjakin) [inline apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js:210]: empty line for readability

File: apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js [hunk 7]
```diff
@@ -188,18 +186,73 @@ class DadataSuggestionProvider extends AbstractSuggestionProvider {
...
+
+        const body = {
+            query,
+            ...this.getContext(context),
+            ...(isNaN(count) ? {} : { count }),
+        }
+
+        if (tin) {
+            const organizationInfo = await this.getOrganization(tin)
+            if (organizationInfo) {
```
Comment (akarjakin) [inline apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js:233]: empty line for readability

File: apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js [hunk 8]
```diff
@@ -188,18 +186,73 @@ class DadataSuggestionProvider extends AbstractSuggestionProvider {
...
+                body.locations_boost = ['settlement_kladr_id', 'city_kladr_id', 'region_kladr_id']
```
Comment (akarjakin) [inline apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js:234]: static array to a constant

File: apps/address-service/domains/common/utils/services/suggest/SuggestionKeystoneApp.js (hunk 1 above)
Comment (AleX83Xpert) [reply apps/address-service/domains/common/utils/services/suggest/SuggestionKeystoneApp.js]: Done

File: apps/address-service/domains/common/utils/services/suggest/SuggestionKeystoneApp.js (hunk 2 above)
Comment (AleX83Xpert) [reply apps/address-service/domains/common/utils/services/suggest/SuggestionKeystoneApp.js]: Done

File: apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js (hunk 4 above)
Comment (AleX83Xpert) [reply apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js:206]: Done

File: apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js (hunk 5 above)
Comment (AleX83Xpert) [reply apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js:219]: Completed

File: apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js (hunk 6 above)
Comment (AleX83Xpert) [reply apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js:215]: ooook, added :)

File: apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js (hunk 7 above)
Comment (AleX83Xpert) [reply apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js:240]: Done

File: apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js (hunk 8 above)
Comment (AleX83Xpert) [reply apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js]: Refactored

File: apps/address-service/domains/common/utils/services/suggest/SuggestionKeystoneApp.js (hunk 3 above)
Comment (AleX83Xpert) [reply apps/address-service/domains/common/utils/services/suggest/SuggestionKeystoneApp.js:66]: Yes, it's ok. I want to get NaN in the case of no parameter. The default value already set within particular provider.

File: apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js [hunk 9]
```diff
@@ -1,8 +1,11 @@
-const { AbstractSuggestionProvider } = require('@address-service/domains/common/utils/services/suggest/AbstractSuggestionProvider')
+const { AbstractSuggestionProvider } = require('@address-service/domains/common/utils/services/suggest/providers/AbstractSuggestionProvider')
 const conf = require('@open-condo/config')
 const get = require('lodash/get')
 const fetch = require('node-fetch')
 const { DADATA_PROVIDER } = require('@address-service/domains/common/constants/providers')
+const { getRedisClient } = require('@open-condo/keystone/redis')
+
+const ORGANIZATION_TIN_CACHE_TTL = 60 // in seconds
```
Comment (dkoviazin) [inline apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js:8]: 1 day would be better (or 1 year)

**Review by dkoviazin — APPROVED** [review]

File: apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js (hunk 9 above)
Comment (AleX83Xpert) [reply apps/address-service/domains/common/utils/services/suggest/providers/DadataSuggestionProvider.js]: Ok, 24h

**Review by akarjakin — APPROVED** [review]

**Pull request merged**